### PR TITLE
Feature/authorization

### PR DESCRIPTION
**Pull request by MagzyCode**

### CODE REVIEW

File: WorkoutGlobal.Api/Controllers/AuthenticationController.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using WorkoutGlobal.Api.Filters.ActionFilters;
+using WorkoutGlobal.Api.Models.DTOs.UserDTOs;
+using WorkoutGlobal.Api.Models.ErrorModels;
+
+namespace WorkoutGlobal.Api.Controllers
+{
+    /// <summary>
+    /// Represents controller for authorization and registration.
+    /// </summary>
+    [Route("api")]
```
Comment (vliashko) [inline WorkoutGlobal.Api/Controllers/AuthenticationController.cs:14]: Not correct global route.

File: WorkoutGlobal.Api/Controllers/AuthenticationController.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        /// <summary>
+        /// Registrate user in system.
+        /// </summary>
+        /// <param name="userRegistrationDto"></param>
+        /// <returns>If user already exists in system return BadRequest status with error model, if don't exists return Created status.</returns>
+        /// <response code="200">User was successfully registered.</response>
+        /// <response code="400">Incoming model already exists in system.</response>
+        /// <response code="500">Something going wrong on server.</response>
+        [HttpPost("registrate")]
```
Comment (vliashko) [inline WorkoutGlobal.Api/Controllers/AuthenticationController.cs:71]: registrate - it's a verb, use 'registration' instead

File: WorkoutGlobal.Api/Controllers/AuthenticationController.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    [Route("api")]
+    [ApiController]
+    [Produces("application/json")]
+    public class AuthenticationController : ControllerBase
+    {
+        private readonly IMapper _mapper;
+        private readonly IRepositoryManager _repositoryManager;
+
+        /// <summary>
+        /// Sets instance of authentication controller with AutoMapper instance.
```
Comment (vliashko) [inline WorkoutGlobal.Api/Controllers/AuthenticationController.cs:23]: Not only AutoMapper reference. It's a ctor, so best way to set summary 'Ctor for ...'. No need to describe hard thing

File: WorkoutGlobal.Api/Extensions/ServiceExtensions.cs [hunk 4]
```diff
@@ -29,7 +35,39 @@ public static class ServiceExtensions
...
+        public static void ConfigureIdentity(this IServiceCollection services)
+        {
+            var builder = services.AddIdentityCore<UserCredentials>(o =>
+            {
+                o.Password.RequireDigit = true;
+                o.Password.RequireLowercase = false;
+                o.Password.RequireUppercase = false;
+                o.Password.RequireNonAlphanumeric = false;
+                o.Password.RequiredLength = 10;
+                o.User.RequireUniqueEmail = true;
```
Comment (vliashko) [inline WorkoutGlobal.Api/Extensions/ServiceExtensions.cs:66]: No need for now to configure email some how.

File: WorkoutGlobal.Api/Migrations/20220506124830_Initial.cs [hunk 5]
```diff
@@ -0,0 +1,243 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore.Migrations;
+
+#nullable disable
+#pragma warning disable 1591
```
Comment (vliashko) [inline WorkoutGlobal.Api/Migrations/20220506124830_Initial.cs:5]: Why we need this? Or it auto-gen?

File: WorkoutGlobal.Api/Models/Validators/UserValidators/UserRegistrationDtoValidator.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    public class UserRegistrationDtoValidator : AbstractValidator<UserRegistrationDto>
+    {
+        /// <summary>
+        /// Sets validation rules for registration user.
+        /// </summary>
+        public UserRegistrationDtoValidator()
+        {
+            RuleFor(user => user.UserName)
+                .NotNull()
+                .NotEmpty()
```
Comment (vliashko) [inline WorkoutGlobal.Api/Models/Validators/UserValidators/UserRegistrationDtoValidator.cs:18]: Should we have NotNull && NotEmpty? If we remove NotNull will it allow nullable?

File: WorkoutGlobal.Api/Models/Validators/UserValidators/UserRegistrationDtoValidator.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                .Length(3, 40)
+                .Matches(@"[^\sА-Яа-я@%?#<>%/]")
```
Comment (vliashko) [inline WorkoutGlobal.Api/Models/Validators/UserValidators/UserRegistrationDtoValidator.cs:20]: Only latin alphabet

File: WorkoutGlobal.Api/Repositories/BaseRepositories/RepositoryManager.cs [hunk 8]
```diff
@@ -9,20 +10,38 @@ namespace WorkoutGlobal.Api.Repositories.BaseRepositories
...
+            IUserCredentialsRepository userCredentialsRepository,
+            IAuthenticationRepository authenticationRepository) 
         {
             _healthRepository = healthRepository;
+            _userCredentialsRepository = userCredentialsRepository;
+            _authenticationRepository = authenticationRepository;
         }
 
         /// <summary>
         /// Health repositoty instance.
```
Comment (vliashko) [inline WorkoutGlobal.Api/Repositories/BaseRepositories/RepositoryManager.cs:33]: 'repositoty' -> 'repository '

File: WorkoutGlobal.UI/ApiConnection/Contracts/IAuthenticationService.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+﻿using Refit;
+using WorkoutGlobal.UI.Models;
+
+namespace WorkoutGlobal.UI.ApiConnection.Contracts
+{
+    /// <summary>
+    /// Represents base structure of authentication endpoints.
+    /// </summary>
+    public interface IAuthenticationService : IApiData
```
Comment (vliashko) [inline WorkoutGlobal.UI/ApiConnection/Contracts/IAuthenticationService.cs:9]: Why we need IAuthenticationService && IApiData? And why interface for contoller with name '...Service'

**Review by vliashko — CHANGES_REQUESTED** [review]

File: WorkoutGlobal.Api/Migrations/20220506124830_Initial.cs (hunk 5 above)
Comment (MagzyCode) [reply WorkoutGlobal.Api/Migrations/20220506124830_Initial.cs:5]: No, it's auto-gen. I'm using it for disable xml-documentation for migrations. I want to use xml-documentation everywhere, except migrations.

File: WorkoutGlobal.Api/Models/Validators/UserValidators/UserRegistrationDtoValidator.cs (hunk 6 above)
Comment (MagzyCode) [reply WorkoutGlobal.Api/Models/Validators/UserValidators/UserRegistrationDtoValidator.cs:18]: No. After reading the documentation on these methods, I made sure that the NotEmpty method also checks for null, so it's worth removing the data chain of the two methods.

File: WorkoutGlobal.UI/ApiConnection/Contracts/IAuthenticationService.cs (hunk 9 above)
Comment (MagzyCode) [reply WorkoutGlobal.UI/ApiConnection/Contracts/IAuthenticationService.cs:9]: I'm using IApiData like limitation in generic class `WorkoutGlobal.UI.ApiConnection.Services.BaseService<T> where T : IApiData`. For me, classes and interfaces that provide access to the api come with the prefix "service". Maybe I'm inaccurate in naming the essence of the class.

File: WorkoutGlobal.UI/Middlewares/GlobalExceptionHandlerMiddleware.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        /// <returns>A task that represents asynchronous operation of going on middleware pipeline.</returns>
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+            }
+            catch (Exception exception)
+            {
+                // TODO: Update this method for existed exceptions.
```
Comment (vliashko) [inline WorkoutGlobal.UI/Middlewares/GlobalExceptionHandlerMiddleware.cs:37]: Need to finish this part.

**Review by vliashko — CHANGES_REQUESTED** [review]

File: WorkoutGlobal.Api.Tests/Controllers/AuthenticationControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            var authenticationRepositoryMock = new Mock<IAuthenticationRepository>();
+            _repositoryManagerMock.Setup(x => x.AuthenticationRepository).Returns(authenticationRepositoryMock.Object);
+            authenticationRepositoryMock.Setup(x => x.IsUserExisted(It.IsAny<UserRegistrationDto>())).Returns(true);
+
+            // act
+            var result = await _authenticationController.Registrate(userRegistrationDto);
+
+            // assert
+            result.Should().NotBeNull()
+                .And.BeOfType<BadRequestObjectResult>()
```
Comment (vliashko) [inline WorkoutGlobal.Api.Tests/Controllers/AuthenticationControllerTests.cs:98]: At this point let put it as is. But in future let use separate part for this. 

result.Should().NotBeNull()
result.Sould().BeOfType<>() 
and so on..

File: WorkoutGlobal.Api.Tests/Repositories/AuthenticationRepositoryTests.cs [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+        public async Task GenerateUserCredentialsAsync_NullUserPassword_ReturnUserPasswordHash()
+        {
+            // act
+            var userCredentialsDto = new UserCredentialsDto()
+            {
+                UserName = null,
+                Email = null,
+                Password = "qwerty123"
+            };
+            _mapper
```
Comment (vliashko) [inline WorkoutGlobal.Api.Tests/Repositories/AuthenticationRepositoryTests.cs:266]: let configure mapper in contructor or initial class.

File: WorkoutGlobal.Api/Repositories/AuthorizationRepositories/AuthenticationRepository.cs [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        /// </summary>
+        /// <param name="userCredentialsDto">User credentials.</param>
+        /// <returns>Existed user.</returns>
+        public UserCredentials FindUserByCredentials(UserCredentialsDto userCredentialsDto)
+        {
+            if (_userManager == null)
+                throw new ArgumentNullException(nameof(userCredentialsDto));
+
+            var userCredentials = Context.UserCredentials
+                .SingleOrDefault(user => user.UserName == userCredentialsDto.UserName);
```
Comment (vliashko) [inline WorkoutGlobal.Api/Repositories/AuthorizationRepositories/AuthenticationRepository.cs:77]: Do not use .SingleOrDefault(user => user.UserName == userCredentialsDto.UserName);

Use .Where(user => user.UserName == userCredentialsDto.UserName).SingleOrDefault()

That can up some memory allocation, but better from timing site

File: WorkoutGlobal.Api/Controllers/AuthenticationController.cs [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        public async Task<IActionResult> Authenticate([FromBody] UserAuthorizationDto userAuthorizationDto)
+        {
+            var isUserValid = await _repositoryManager.AuthenticationRepository.ValidateUserAsync(userAuthorizationDto);
+
+            if (!isUserValid)
+                return Unauthorized(new ErrorDetails()
+                {
+                    StatusCode = StatusCodes.Status401Unauthorized,
+                    Message = "Authenticate was failed. Try another user name or password",
+                    Details = new StackTrace().ToString()
```
[the remark this answers is not in the document]
Comment (MagzyCode) [reply WorkoutGlobal.Api/Controllers/AuthenticationController.cs:55]: Common stack trace of call.

**Review by vliashko — APPROVED** [review]
Looks good!

**Pull request merged**